apps/ws/src/socket/ws.gateway.ts

```diff
@@ -1,41 +1,113 @@
+const nr = require('newrelic');
+
+import { OnGatewayConnection, OnGatewayDisconnect, WebSocketGateway, WebSocketServer } from '@nestjs/websockets';
 import { Server, Socket } from 'socket.io';
 import { JwtService } from '@nestjs/jwt';
-import { ISubscriberJwt } from '@novu/shared';
+import { ISubscriberJwt, ObservabilityBackgroundTransactionEnum } from '@novu/shared';
 import { SubscriberOnlineService } from '../shared/subscriber-online';
 
 @WebSocketGateway()
-export class WSGateway implements OnGatewayConnection {
+export class WSGateway implements OnGatewayConnection, OnGatewayDisconnect {
   constructor(private jwtService: JwtService, private subscriberOnlineService: SubscriberOnlineService) {}
 
   @WebSocketServer()
   server: Server;
 
+  async handleDisconnect(connection: Socket) {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const _this = this;
+
+    return new Promise((resolve, reject) => {
+      nr.startBackgroundTransaction(
+        ObservabilityBackgroundTransactionEnum.WS_HANDLE_SOCKET_DISCONNECT,
+        'WS Service',
+        function () {
+          const transaction = nr.getTransaction();
+
+          _this
+            .processDisconnectionRequest(connection)
+            .then(resolve)
+            .catch(reject)
+            .finally(() => {
+              transaction.end();
+            });
+        }
+      );
+    });
+  }
+
   async handleConnection(connection: Socket) {
-    const token = connection.handshake.auth?.token || connection.handshake.query?.token;
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const _this = this;
 
-    if (!token || token === 'null') {
-      return this.disconnect(connection);
-    }
+    return new Promise((resolve, reject) => {
+      nr.startBackgroundTransaction(
+        ObservabilityBackgroundTransactionEnum.WS_HANDLE_SOCKET_CONNECTION,
+        'WS Service',
+        function () {
+          const transaction = nr.getTransaction();
 
+          _this
+            .processConnectionRequest(connection)
+            .then(resolve)
+            .catch(reject)
+            .finally(() => {
+              transaction.end();
+            });
+        }
+      );
+    });
+  }
+
+  private extractToken(connection: Socket): string | undefined {
+    return connection.handshake.auth?.token || connection.handshake.query?.token;
+  }
+
+  private async getSubscriber(token: string): Promise<ISubscriberJwt | undefined> {
     let subscriber: ISubscriberJwt;
 
     try {
       subscriber = await this.jwtService.verify(token as string);
       if (subscriber.aud !== 'widget_user') {
-        return this.disconnect(connection);
+        return;
       }
+
+      return subscriber;
     } catch (e) {
+      return;
+    }
+  }
+
+  private async processDisconnectionRequest(connection: Socket) {
+    const token = this.extractToken(connection);
+
+    if (!token || token === 'null') {
+      return;
```

Comment:
should we try to disconnect here and on the 'return' below??
```suggestion
      return this.disconnect(connection);
 ```

Reply:
This hook is fired on the disconnect action, so it will disconnect regardless here